### PR TITLE
Hl/improve UI table

### PR DESCRIPTION
**Pull request by hussam789**
## **Type**
Enhancement

___

## **Description**
This PR primarily focuses on enhancing the presentation of code suggestions. The key changes include:
- A new function `replace_code_tags` in `pr_agent/algo/utils.py` that replaces instances of backticks (`) with `<code>` and `</code>` tags.
- Significant refactoring of the `publish_summarizes_suggestions` method in `pr_agent/tools/pr_code_suggestions.py` to organize suggestions by label and present them in a more structured and readable format.
- The function `_insert_br_after_x_chars` in `pr_agent/tools/pr_description.py` has been renamed to `insert_br_after_x_chars` and moved to the global scope. This function inserts a line break into a string after a word that increases its length above a certain number of characters.
- The Pydantic definition for `CodeSuggestion` in `pr_agent/settings/pr_code_suggestions_prompts.toml` has been updated with a new field `one_sentence_summary`. The description for the `existing_code` and `improved_code` fields have also been updated.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th>Relevant files&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>utils.py
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        pr_agent/algo/utils.py<br><br>

**A new function `replace_code_tags` has been added. This <br>function replaces odd instances of backticks (`) with <br>`<code>` and even instances of backticks with `</code>`.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/597/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53"> </a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_code_suggestions.py
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        pr_agent/tools/pr_code_suggestions.py<br><br>

**The `publish_summarizes_suggestions` method has been <br>significantly refactored. It now organizes suggestions by <br>label and presents them in a more structured and readable <br>format. The method `run` has also been updated with <br>commented-out code for testing purposes. The `difflib` <br>library has been imported for generating unified diffs.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/597/files#diff-b57ba775e741d6f80bc4f8154b71330c011dae0ac43f3d0197e785b3e6b7117b"> </a></td>

</tr>                    
</table></td></tr><tr><td><strong>Refactoring
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pr_description.py
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        pr_agent/tools/pr_description.py<br><br>

**The function `_insert_br_after_x_chars` has been renamed to <br>`insert_br_after_x_chars` and moved to the global scope. <br>This function inserts a line break into a string after a <br>word that increases its length above a certain number of <br>characters.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/597/files#diff-66130451adce278a098a5770e2db52b12050205fcd4d80c13f813615d4449abb"> </a></td>

</tr>                    
</table></td></tr><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pr_code_suggestions_prompts.toml
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        pr_agent/settings/pr_code_suggestions_prompts.toml<br><br>

**The Pydantic definition for `CodeSuggestion` has been <br>updated with a new field `one_sentence_summary`. The <br>description for the `existing_code` and `improved_code` <br>fields have also been updated.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/597/files#diff-81979f64eddc8f93a5c6ea83ff1e7ba92e551acaecd310c606255161e914c2ea"> </a></td>

</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: pr_agent/algo/utils.py [hunk 1]
```diff
@@ -474,4 +474,13 @@ def clip_tokens(text: str, max_tokens: int, add_three_dots=True) -> str:
...
         get_logger().warning(f"Failed to clip tokens: {e}")
-        return text
\ No newline at end of file
+        return text
+
+def replace_code_tags(text):
+    """
+    Replace odd instances of ` with <code> and even instances of ` with </code>
+    """
+    parts = text.split('`')
```
Comment (mrT23) [inline pr_agent/algo/utils.py:483]: i would add a check first to see if '`' is in the original text, and if not, return the input

File: pr_agent/tools/pr_code_suggestions.py [hunk 2]
```diff
@@ -55,6 +56,8 @@ def __init__(self, pr_url: str, cli_mode=False, args: list = None,
...
...ing_code': 'id: description2\n', 'improved_code': 'id: additional_details\n', 'relevant_lines_start': 13, 'relevant_lines_end': 13, 'label': 'enhancement'}, {'relevant_file': '.github/ISSUE_TEMPLATE/sweep-feature.yml', 'suggestion_content': 'The `description` attribute for the textarea is "More details for Sweep", which is not very descriptive. Consider providing a more specific description.\n', 'existing_code': 'description: More details for Sweep\n', 'improved_code': 'description: Please provide more details about the feature you want to test.\n', 'relevant_lines_start': 10, 'relevant_lines_end': 10, 'label': 'enhancement'}, {'relevant_file': '.github/ISSUE_TEMPLATE/sweep-refactor.yml', 'suggestion_content': 'The `placeholder` attribute for the textarea is "We are migrating this function to ... version because ...", which is not very clear. Consider providing a more specific placeholder.\n', 'existing_code': 'placeholder: We are migrating this function to ... version because ...\n', 'improved_code': 'placeholder: We are migrating this function to version 2.0 because it provides better performance.\n', 'relevant_lines_start': 11, 'relevant_lines_end': 11, 'label': 'enhancement'}]}
```
Comment (mrT23) [inline pr_agent/tools/pr_code_suggestions.py:59]: remove this

File: pr_agent/settings/pr_code_suggestions_prompts.toml [hunk 3]
```diff
@@ -47,14 +47,15 @@ Extra instructions from the user:
...
 =====
 class CodeSuggestion(BaseModel):
     relevant_file: str = Field(description="the relevant file full path")
     suggestion_content: str = Field(description="an actionable suggestion for meaningfully improving the new code introduced in the PR")
 {%- if summarize_mode %}
-    existing_code: str = Field(description="a short code snippet from a '__new hunk__' section to illustrate the relevant existing code. Don't show the line numbers. Shorten parts of the code ('...') if needed")
-    improved_code: str = Field(description="a short code snippet to illustrate the improved code, after applying the suggestion. Shorten parts of the code ('...') if needed")
+    existing_code: str = Field(description="a short code snippet from a '__new hunk__' section to illustrate the relevant existing code. Don't show the line numbers.")
+    improved_code: str = Field(description="a short code snippet to illustrate the improved code, after applying the suggestion.")
+    one_sentence_summary:str = Field(description="a short summary of the suggestion action, in a single sentence. Focus on the 'what'. Be general, and avoid method or variable names.")
```
Comment (mrT23) [inline pr_agent/settings/pr_code_suggestions_prompts.toml:58]: lets try to sharpen even further this, to avoid the 'in method'

**Review by mrT23 — APPROVED** [review]

**Review by barnett-yuxiang — COMMENTED** [review]
PTAL

**Comment by hussam789** [comment]
Preparing PR description...

**Comment by hussam789** [comment]
Preparing PR description...

**Comment by hussam789** [comment]
Preparing PR description...

File: pr_agent/algo/utils.py
Comment (hussam789) [inline pr_agent/algo/utils.py:1]: <details>
      <summary><strong>File changes summary</strong></summary>
    <ul>

A new function `replace_code_tags` has been added. This <br>function replaces odd instances of backticks (`) with <br>`<code>` and even instances of backticks (`) with `</code>`.
    </ul>
</details>

File: pr_agent/tools/pr_code_suggestions.py
Comment (hussam789) [inline pr_agent/tools/pr_code_suggestions.py:1]: <details>
      <summary><strong>File changes summary</strong></summary>
    <ul>

Major changes in the `publish_summarizes_suggestions` <br>function. The function now generates a more structured and <br>detailed HTML table for PR suggestions. It also uses the <br>`difflib` library to generate a unified diff between the <br>existing and improved code. The `replace_code_tags` function <br>from `pr_agent.algo.utils` is used to handle backticks in <br>the suggestion summary.
    </ul>
</details>

File: pr_agent/tools/pr_description.py
Comment (hussam789) [inline pr_agent/tools/pr_description.py:1]: <details>
      <summary><strong>File changes summary</strong></summary>
    <ul>

The function `_insert_br_after_x_chars` has been renamed to <br>`insert_br_after_x_chars` and moved to the global scope. <br>This function inserts a `<br>` tag into a string after a <br>word that increases its length above a certain number of <br>characters.
    </ul>
</details>

File: pr_agent/settings/pr_code_suggestions_prompts.toml
Comment (hussam789) [inline pr_agent/settings/pr_code_suggestions_prompts.toml:1]: <details>
      <summary><strong>File changes summary</strong></summary>
    <ul>

The Pydantic definitions for the PRCodeSuggestions object <br>have been updated. A new field `one_sentence_summary` has <br>been added to the `CodeSuggestion` class. The descriptions <br>for `existing_code` and `improved_code` have also been <br>updated.
    </ul>
</details>

**Comment by hussam789** [comment]
**[PR Description](https://github.com/Codium-ai/pr-agent/pull/597)** updated to latest commit (https://github.com/Codium-ai/pr-agent/commit/d7e0aad5276c0e2a0b2b7f3580119fed4929abeb)

**Pull request merged**